pkg/runtime.v2/core/trainingruntime.go

```diff
@@ -0,0 +1,121 @@
+	fwkplugins "github.com/kubeflow/training-operator/pkg/runtime.v2/framework/plugins"
+	idxer "github.com/kubeflow/training-operator/pkg/runtime.v2/indexer"
+)
+
+var (
+	errorNotFoundSpecifiedTrainingRuntime = errors.New("not found TrainingRuntime specified in TrainJob")
+)
+
+type TrainingRuntime struct {
+	framework *fwkcore.Framework
+	client    client.Client
+	scheme    *apiruntime.Scheme
+}
+
+var TrainingRuntimeGroupKind = schema.GroupKind{
+	Group: kubeflowv2.GroupVersion.Group,
+	Kind:  "TrainingRuntime",
+}.String()
+
+var _ runtime.Runtime = (*TrainingRuntime)(nil)
+
+var trainingRuntimeFactory *TrainingRuntime
+
+func NewTrainingRuntime(ctx context.Context, c client.Client, indexer client.FieldIndexer) (runtime.Runtime, error) {
+	if err := indexer.IndexField(ctx, &kubeflowv2.TrainJob{}, idxer.TrainJobTrainingRuntimeRefKey, idxer.IndexTrainJobTrainingRuntimes); err != nil {
+		return nil, fmt.Errorf("setting index on TrainingRuntime and ClusterTrainigRuntime for TrainJob: %w", err)
+	}
+	fwk, err := fwkcore.New(ctx, c, fwkplugins.NewRegistry(), indexer)
+	if err != nil {
+		return nil, err
+	}
+	trainingRuntimeFactory = &TrainingRuntime{
+		framework: fwk,
+		client:    c,
+		scheme:    c.Scheme(),
+	}
+	return trainingRuntimeFactory, nil
+}
+
+func (r *TrainingRuntime) NewObjects(ctx context.Context, trainJob *kubeflowv2.TrainJob) ([]client.Object, error) {
+	var trainingRuntime kubeflowv2.TrainingRuntime
+	err := r.client.Get(ctx, client.ObjectKey{Namespace: trainJob.Namespace, Name: trainJob.Spec.TrainingRuntimeRef.Name}, &trainingRuntime)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", errorNotFoundSpecifiedTrainingRuntime, err)
+	}
+	return r.buildObjects(ctx, trainJob, trainingRuntime.Spec.Template, trainingRuntime.Spec.MLPolicy, trainingRuntime.Spec.PodGroupPolicy)
+}
+
+func (r *TrainingRuntime) buildObjects(ctx context.Context, trainJob *kubeflowv2.TrainJob, jobSetTemplateSpec kubeflowv2.JobSetTemplateSpec,
+	mlPolicy *kubeflowv2.MLPolicy, podGroupPolicy *kubeflowv2.PodGroupPolicy) ([]client.Object, error) {
+	opts := []runtime.InfoOption{
+		runtime.WithLabels(jobSetTemplateSpec.Labels),
+		runtime.WithAnnotations(jobSetTemplateSpec.Annotations),
+		runtime.WithMLPolicy(mlPolicy),
+		runtime.WithPodGroupPolicy(podGroupPolicy),
+	}
+	for idx, rJob := range jobSetTemplateSpec.Spec.ReplicatedJobs {
+		if rJob.Replicas == 0 {
+			jobSetTemplateSpec.Spec.ReplicatedJobs[idx].Replicas = 1
+		}
+		replicas := jobSetTemplateSpec.Spec.ReplicatedJobs[idx].Replicas * ptr.Deref(rJob.Template.Spec.Completions, 1)
+		opts = append(opts, runtime.WithPodSpecReplicas(rJob.Name, replicas, rJob.Template.Spec.Template.Spec))
+	}
+	info := runtime.NewInfo(&jobsetv1alpha2.JobSet{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: jobsetv1alpha2.SchemeGroupVersion.String(),
+			Kind:       "JobSet",
+		},
+		Spec: *jobSetTemplateSpec.Spec.DeepCopy(),
+	}, opts...)
+
+	if err := r.framework.RunEnforceMLPolicyPlugins(info); err != nil {
```

Comment:
What about DatasetConfig, ModelConfig, and PodSpecOverrides, how will enforce them on JobSet template ?

Reply:
I can imagine implementing `EnforceDatasetConfigPlugins` extension points as well as `EnforceMLPolicyPlugins`.
I'm happy to implement the interface after this PR.